File: backend/src/main/java/backend/resource/ToDoResource.java
package backend.resource;

import java.util.Optional;

import javax.transaction.Transactional;
import javax.ws.rs.BeanParam;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import backend.helper.BeanHelper;
import backend.repository.ToDoRepository;
import backend.repository.entity.ToDo;
import backend.resource.bean.ListParam;
import backend.resource.bean.ToDoBean;
import backend.resource.bean.ToDoListBean;
import backend.resource.bean.ToDoBean.Category;
import backend.resource.exception.NotFoundException;

/** A JAX-RS (Jersey) resource class for ToDo items */
@Component
@Path("todos")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
@Transactional
public class ToDoResource {

	@Autowired private ToDoRepository repository;
	@Autowired private BeanHelper beanHelper;

	/** GET /api/todos: Get a list of todo items. */
	@GET
	public Response getList(@BeanParam ListParam listParam) {
		Pageable pageable = PageRequest.of(listParam.getPage(), listParam.getSize());
		Page<ToDo> entityList = repository.findAll(pageable);
		ToDoListBean list = new ToDoListBean();
		list.setItems(beanHelper.createAndCopyIterable(entityList, ToDoBean.class, (ToDo src, ToDoBean target) -> {
			target.setCategory(Category.valueOf(src.getCategory().toString()));
		}));
		list.setTotal(entityList.getTotalElements());
		return Response.ok().entity(list).build();
	}

	/** POST /api/todos: Create a new todo item. */
	@POST
	public Response create(ToDoBean reqBean) {
		ToDo entity = beanHelper.createAndCopy(reqBean, ToDo.class, (ToDoBean src, ToDo target) -> {
			target.setId(null);
			target.setCategory(ToDo.Category.valueOf(src.getCategory().toString()));
		});
		entity = repository.save(entity);
		ToDoBean resBean = beanHelper.createAndCopy(entity, ToDoBean.class, (ToDo src, ToDoBean target) -> {
			target.setCategory(Category.valueOf(src.getCategory().toString()));
		});
		return Response.ok().entity(resBean).build();
	}

	/** GET /api/todos/{id}: Get a todo item. */
	@GET @Path("{id}")
	public Response get(@PathParam("id") Long id) {
		Optional<ToDo> result = repository.findById(id);
		if(result.isPresent()) {
			ToDoBean bean = beanHelper.createAndCopy(result.get(), ToDoBean.class, (ToDo src, ToDoBean target) -> {
				target.setCategory(Category.valueOf(src.getCategory().toString()));
			});
			return Response.ok().entity(bean).build();
		} else {
			throw new NotFoundException();
		}
	}

	/** PUT /api/todos/{id}: Update a todo item. */
	@PUT @Path("{id}")
	public Response update(@PathParam("id") Long id, ToDoBean reqBean) {
		ToDo entity = beanHelper.createAndCopy(reqBean, ToDo.class, (ToDoBean src, ToDo target) -> {
			target.setId(id);
			target.setCategory(ToDo.Category.valueOf(src.getCategory().toString()));
		});
		repository.save(entity);
		return Response.ok().entity(reqBean).build();
	}

	/** DELETE /api/todos/{id}: Delete a todo item. */
	@DELETE @Path("{id}")
	public Response remove(@PathParam("id") Long id) {
		Optional<ToDo> result = repository.findById(id);
		if(result.isPresent()) {
			repository.deleteById(id);
			return Response.ok().build();
		} else {
			throw new NotFoundException();
		}
	}
}

File: backend/src/main/kotlin/backend/resource/ToDoResource.kt
package backend.resource

import backend.helper.BeanHelper
import backend.repository.ToDoRepository
import backend.repository.entity.ToDo
import backend.resource.bean.ListParam
import backend.resource.bean.ToDoBean
import backend.resource.bean.ToDoListBean
import backend.resource.exception.NotFoundException
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import javax.transaction.Transactional
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/** A JAX-RS (Jersey) resource class for ToDo items  */
// Resource classes must be annotated with one of @Component, @Service, @Controller, @Repository.
// (Note: Spring AOP doesn't work if @Named is used instead.)
@Component
@Path("todos") // Base path for all methods in the class
@Consumes(MediaType.APPLICATION_JSON) // Content-Type of request body
@Produces(MediaType.APPLICATION_JSON) // Content-Type of response body
@Transactional // Make all the methods in the class transactional.
class ToDoResource (private val repository: ToDoRepository, val beanHelper: BeanHelper) {
    val beanToEntity = { src: ToDoBean, target: ToDo ->
        target.category = ToDo.Category.valueOf(src.category.toString())
    }

    val entityToBean = { src: ToDo, target: ToDoBean ->
        target.category = ToDoBean.Category.valueOf(src.category.toString()) }

    /** GET /api/todos: Get a list of todo items.  */
    @GET
    fun getList(@BeanParam listParam: ListParam): Response {
        val pageable: Pageable = PageRequest.of(listParam.page, listParam.size)
        val entityList = repository.findAll(pageable)
        var list = ToDoListBean()
        list.items = beanHelper.createAndCopyIterable(entityList, ToDoBean::class.java, entityToBean)
        list.total = entityList.totalElements
        return Response.ok().entity(list).build()
    }

    /** POST /api/todos: Create a new todo item.  */
    @POST
    fun create(reqBean: ToDoBean): Response {
        var entity = beanHelper.createAndCopy(reqBean, ToDo::class.java, beanToEntity)
        entity = repository.save(entity)
        val resBean = beanHelper.createAndCopy(entity, ToDoBean::class.java, entityToBean)
        return Response.ok().entity(resBean).build()
    }

    /** GET /api/todos/{id}: Get a todo item.  */
    @GET
    @Path("{id}")
    operator fun get(@PathParam("id") id: Long): Response {
        val result = repository.findById(id)
        return if (result.isPresent) {
            val bean = beanHelper.createAndCopy(result.get(), ToDoBean::class.java, entityToBean)
            Response.ok().entity(bean).build()
        } else {
            throw NotFoundException()
        }
    }

    /** PUT /api/todos/{id}: Update a todo item.  */
    @PUT
    @Path("{id}")
    fun update(@PathParam("id") id: Long?, reqBean: ToDoBean): Response {
        reqBean.id = id
        val entity = beanHelper.createAndCopy(reqBean, ToDo::class.java, beanToEntity)
        repository.save(entity)
        return Response.ok().entity(reqBean).build()
    }

    /** DELETE /api/todos/{id}: Delete a todo item.  */
    @DELETE
    @Path("{id}")
    fun remove(@PathParam("id") id: Long): Response {
        val result = repository.findById(id)
        return if (result.isPresent) {
            repository.deleteById(id)
            Response.ok().build()
        } else {
            throw NotFoundException()
        }
    }
}